### PR TITLE
[feat]: [CDS-58165]: add facade for merge input apis

### PR DESCRIPTION
**Pull request by tchaurasiya**
## Describe your changes

Added Facade to absract out service override utilities at execution time.
Refactoring of mudules
Only classes that has relevant changes are following, rest changes are for refactoring
125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceOverrideUtilityFacade.java
125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/beans/ServiceStepV3Parameters.java
125-cd-nextgen/src/test/java/io/harness/cdng/service/steps/helpers/ServiceOverrideUtilityFacadeTest.java

Main Functionalities Details:
**mergeServiceOverrideInputs**
- based on isV2 field of override, apply merging logic
- if isV2 - false - same old merging taka place with override yaml
- isV2 - true - new logic compatible with spec takes place
- Assumption is all overrides will have isV2 field set. (as on demand migration is already in place)

**mergeEnvironmentInputs**
- based on environment - isV2 field, above explained logic is implemented
- Next ToDo - add project setting logic for the same

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/48088)

### CODE REVIEW

**Comment by tchaurasiya** [comment]
trigger review

**Comment by tchaurasiya** [comment]
trigger uts

**Comment by tchaurasiya** [comment]
trigger uts

**Comment by yogesh-chauhan** [comment]
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceOverrideUtilityFacade.java` line 77 at r1](https://reviewable.io/reviews/harness/harness-core/48088#-NW2GBXGEl0dv11ozfo1:-NW2GBXGEl0dv11ozfo2:b-6xtyn8) ([raw file](https://github.com/harness/harness-core/blob/d0f0d4a40c05ad4d75037721417664489cfad252/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceOverrideUtilityFacade.java#L77)):*
> ```Java
> 
>   public NGServiceOverrideConfigV2 mergeServiceOverridesInputs(
>       NGServiceOverridesEntity entity, ParameterField<Map<String, Object>> serviceOverrideInputs) {
> ```

if we are passing parameter field here, we should take care of it being an expression also. Can we see how we can just pass the Map here ?

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceOverrideUtilityFacade.java [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+    Map<String, Object> inputs = new HashMap<>();
+    inputs.put(YamlTypes.SERVICE_OVERRIDE, serviceOverrideInputs);
+    return inputs;
+  }
+
+  private static String getSpecYamlForMerging(NGServiceOverridesEntity entity) throws IOException {
+    String specYaml = CDYamlUtils.getYamlString(entity.getSpec());
+    YamlField yamlField = YamlUtils.readTree(specYaml);
+    JsonNode currJsonNode = yamlField.getNode().getCurrJsonNode();
+    ObjectMapper mapper = new ObjectMapper();
```
Comment (yogesh-chauhan) [inline 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceOverrideUtilityFacade.java:293]: re-use mapper

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceOverrideUtilityFacade.java [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+                                    .type(envGlobalOverrideEntity.getType())
+                                    .spec(specWithMergedInput)
+                                    .build();
+    return ngServiceOverrideConfigV2;
+  }
+
+  private NGServiceOverrideConfigV2 getOverrideConfigV2ForNoInputs(
+      NGServiceOverridesEntity envGlobalOverrideEntity, @NonNull Environment envEntity) throws IOException {
+    NGServiceOverrideConfigV2 ngServiceOverrideConfigV2;
+    if (Boolean.TRUE.equals(envEntity.getIsMigratedToOverride()) || envGlobalOverrideEntity != null) {
```
Comment (yogesh-chauhan) [inline 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceOverrideUtilityFacade.java:204]: can we make sure that it does not fail for someone who has not created any overrides

**Review by yogesh-chauhan — APPROVED** [review]

**Review by yogesh-chauhan — COMMENTED** [review]
> Reviewed 48 of 49 files at r1, 9 of 9 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48088)* status: 50 file(s) to review. 1 of 2 approvals obtained.

**Review by yogesh-chauhan — DISMISSED** [review]
> Reviewed 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48088)* status: 50 file(s) to review. 1 of 2 approvals obtained.

**Comment by tchaurasiya** [comment]
trigger uts

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 3 of 3 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48088)* status: 50 file(s) to review. 1 of 2 approvals obtained.

**Comment by tchaurasiya** [comment]
trigger uts

**Review by yogesh-chauhan — APPROVED** [review]

**Review by vaibhav-si — COMMENTED** [review]
I cannot review such big PR. Is it not possible to break it down into smaller chunks?

Like the movement of code can be one PR. DTOs can be another. Logical change can be another. Just examples, I don't have the full context here.

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48088#-:-NW5q35C5XRsyFU9XB0B:b-nzdunj)* status: 50 file(s) to review. 1 of 2 approvals obtained.

**Comment by tchaurasiya** [comment]
> I cannot review such big PR. Is it not possible to break it down into smaller chunks?
> 
> Like the movement of code can be one PR. DTOs can be another. Logical change can be another. Just examples, I don't have the full context here.
> 
> > _[Reviewable](https://reviewable.io/reviews/harness/harness-core/48088#-:-NW5q35C5XRsyFU9XB0B:b-nzdunj)_ status: 50 file(s) to review. 1 of 2 approvals obtained.

Yes Agree @vaibhav-si.
But In this case, i was not sure of refactoring part, at the start. So need to to do mid way.
Only classes that has relevant changes are 
125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/ServiceOverrideUtilityFacade.java
125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/helpers/beans/ServiceStepV3Parameters.java
125-cd-nextgen/src/test/java/io/harness/cdng/service/steps/helpers/ServiceOverrideUtilityFacadeTest.java

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48088)* status: 50 file(s) to review. 1 of 2 approvals obtained.

**Comment by tchaurasiya** [comment]
trigger review

**Pull request merged**